### PR TITLE
fix(sdk): Allow other nvidia gpu resource types. Fixes #5026

### LINKED ISSUE #5026: Nvidia A100: GPU Slice definition not supported via Pipeline DSL
### What steps did you take:
I created a simple test pipeline to check the Nvidia slicing definitions, using following code
```
train = dsl.ContainerOp(
		name='Train Model',
		image='tensorflow:v1',
	).add_node_selector_constraint("nvidia.com/gpu.product", "A100-SXM4-40GB MIG 1g.5gb")
	train.set_gpu_limit(1)
```

We have 3 node A100 setup on which the Kubeflow version 1.1 deployed and we are following [Mixed strategy ], (https://developer.nvidia.com/blog/getting-kubernetes-ready-for-the-a100-gpu-with-multi-instance-gpu/) basically the MiG mode is enabled. 

### What happened:

Compiling and deploying the YAML definition, Pipeline run is showing an error, please see the attached screenshot
kubeflowerrormig

Changing the DSL definition 
```
.add_node_selector_constraint("nvidia.com/gpu.product", "A100-SXM4-40GB MIG 1g.5gb")
trainOp.set_gpu_limit(1)

```
Creates following error, where the node is not able to be scheduled, (this may work if we dont create the MiG definition)
a100error2

Then changing to last third option where the DSL , compilation error gets thrown, which basically complains about the vendor name to be Nvidia

```
.set_gpu_limit(1, "nvidia.com/mig-3g.20gb")
train.add_node_selector_constraint("nvidia.com/gpu.product","A100-SXM4-40GB")


File "D:\Apps\Anaconda3\lib\site-packages\kfp\dsl\_container_op.py", line 303, in set_gpu_limit
    raise ValueError('vendor can only be nvidia or amd.')
ValueError: vendor can only be nvidia or amd.

```

When I changed the compiled YAML file to change I could run the pipeline successfully.

```
 resources:
        limits:
          nvidia.com/mig-3g.20gb: 1
```

### Environment:

How did you deploy Kubeflow Pipelines (KFP)?

KFP version:  1.0

KFP SDK version:  1.0.4

### Anything else you would like to add:
[Miscellaneous information that will assist in solving the issue.]

/kind bug

/area sdk

### PR DESCRIPTION
**Pull request by mathematiguy**
**Description of your changes:**

- Allows more general NVIDIA GPU setups by permitting resources to take on a wider set of values
- Compatible with the new A100 servers, which use the resource names `nvidia.com/mig-1g.10gb`, `nvidia.com/mig-2g.20gb`, `nvidia.com/mig-4g.40gb` and `nvidia.com/mig-7g.80gb`.
- I also note in #5026 that @kd303's A100 GPU has the resource name `nvidia.com/mig-3g.20gb`. This change would work for his setup also.

I could have fixed it to specifically those resource names, but I imagine there may be other setups out there I am not aware of and opted to accept any resource name starting with `nvidia.com/`

I have tested this change on my A100 server setup and it works fine.

### CODE REVIEW

**Comment by mathematiguy** [comment]
> Thanks for your pull request. It looks like this may be your first contribution to a Google open source project (if not, look below for help). Before we can look at your pull request, you'll need to sign a Contributor License Agreement (CLA).
> 
> memo **Please visit https://cla.developers.google.com/ to sign.**
> 
> Once you've signed (or fixed any issues), please reply here with `@googlebot I signed it!` and we'll verify it.
> #### What to do if you already signed the CLA
> ##### Individual signers
> 
>     * It's possible we don't have your GitHub username or you're using a different email address on your commit. Check [your existing CLA data](https://cla.developers.google.com/clas) and verify that your [email is set on your git commits](https://help.github.com/articles/setting-your-email-in-git/).
> 
> 
> ##### Corporate signers
> 
>     * Your company has a Point of Contact who decides which employees are authorized to participate. Ask your POC to be added to the group of authorized contributors. If you don't know who your Point of Contact is, direct the Google project maintainer to [go/cla#troubleshoot](http://go/cla#troubleshoot) ([Public version](https://opensource.google/docs/cla/#troubleshoot)).
> 
>     * The email used to register you as an authorized contributor must be the email used for the Git commit. Check [your existing CLA data](https://cla.developers.google.com/clas) and verify that your [email is set on your git commits](https://help.github.com/articles/setting-your-email-in-git/).
> 
>     * The email used to register you as an authorized contributor must also be [attached to your GitHub account](https://github.com/settings/emails).
> 
> 
> information_source **Googlers: [Go here](https://goto.google.com/prinfo/https%3A%2F%2Fgithub.com%2Fkubeflow%2Fpipelines%2Fpull%2F6560) for more info**.

@googlebot I signed it!

**Comment by chensun** [comment]
@mathematiguy Thank you for your PR. Can you please add an entry in the release note under "Current version": https://github.com/kubeflow/pipelines/blob/master/sdk/RELEASE.md#current-version-still-in-development

File: sdk/python/kfp/compiler/_op_to_template.py [hunk 1]
```diff
@@ -315,7 +315,9 @@ def _op_to_template(op: BaseOp):
     if isinstance(op, dsl.ContainerOp) and ('resources' in op.container.keys()):
         for setting, val in op.container['resources'].items():
             for resource, param in val.items():
-                if (resource in ['cpu', 'memory', 'amd.com/gpu', 'nvidia.com/gpu'] or re.match('^{{inputs.parameters.*}}$', resource))\
+                if (resource in ['cpu', 'memory', 'amd.com/gpu'] or\
+                    resource.startswith('nvidia.com/') or\
+                    or re.match('^{{inputs.parameters.*}}$', resource))\
```
Comment (chensun) [inline sdk/python/kfp/compiler/_op_to_template.py:320]: you have double "or" here

**Comment by rickysage** [comment]
Hi @mathematiguy and @chensun. Thanks for working this out. When is this PR expected to be merged?

**Pull request closed without merging**